example/contrib/socks-client/socks/impl/handshake.hpp

```diff
@@ -0,0 +1,598 @@
+        , stream_(stream)
+        , hostname_(hostname)
+        , port_(port)
+        , username_(username)
+    {
+        (*this)({}, 0); // start the operation
+    }
+
+    void
+    operator()(
+        error_code ec,
+        std::size_t bytes_transferred)
+    {
+        boost::ignore_unused(bytes_transferred);
+
+        using detail::write;
+        using detail::read;
+
+        auto& response = *response_;
+        auto& request = *request_;
+
+        switch (ec ? 3 : step_)
+        {
+        case 0:
+        {
+            step_ = 1;
+
+            std::size_t bytes_to_write = 9 + username_.size();
+            auto req = static_cast<char*>(request.prepare(bytes_to_write).data());
+
+            write<uint8_t>(detail::SOCKS_VERSION_4, req); // SOCKS VERSION 4.
+            write<uint8_t>(detail::SOCKS_CMD_CONNECT, req); // CONNECT.
+
+            write<uint16_t>(port_, req); // DST PORT.
+
+            auto address = net::ip::make_address_v4(hostname_, ec);
+            if (ec)
+                break;
+
+            write<uint32_t>(address.to_uint(), req); // DST I
+
+            if (!username_.empty())
+            {
+                std::copy(username_.begin(), username_.end(), req);    // USERID
+                req += username_.size();
+            }
+            write<uint8_t>(0, req); // NULL.
+
+            request.commit(bytes_to_write);
+            return net::async_write(stream_, request, std::move(*this));
+        }
+        case 1:
+        {
+            step_ = 2;
+            return net::async_read(stream_, response,
+                net::transfer_exactly(8), std::move(*this));
+        }
+        case 2:
+        {
+            auto resp = static_cast<const unsigned char*>(response.data().data());
+
+            read<uint8_t>(resp); // VN is the version of the reply code and should be 0.
+            auto cd = read<uint8_t>(resp);
+
+            if (cd != detail::SOCKS4_REQUEST_GRANTED)
+            {
+                switch (cd)
+                {
+                case detail::SOCKS4_REQUEST_REJECTED_OR_FAILED:
+                    ec = error::socks_request_rejected_or_failed;
+                    break;
+                case detail::SOCKS4_CANNOT_CONNECT_TARGET_SERVER:
+                    ec = error::socks_request_rejected_cannot_connect;
+                    break;
+                case detail::SOCKS4_REQUEST_REJECTED_USER_NO_ALLOW:
+                    ec = error::socks_request_rejected_incorrect_userid;
+                    break;
+                default:
+                    ec = error::socks_unknown_error;
+                    break;
+                }
+            }
+            BOOST_FALLTHROUGH;
+        }
+        case 3:
+            break;
+        }
+
+        this->complete_now(ec);
+    }
+
+private:
+    Stream& stream_;
+
+    using BufferPtr = std::unique_ptr<Buffer>;
+    BufferPtr request_{ new Buffer() }; // std::make_unique c++14 or later.
+    BufferPtr response_{ new Buffer() };
```

Comment:
I don't think it is necessary to have two buffers. After sending the request, we can free the memory and then reallocate. Or perhaps we can recycle the one buffer - the smallest request is 9 bytes, and it looks like the response is always 8 bytes, so we can get away with just storing one `std::unique_ptr<char[], my_deleter>`